url29ecd5d728_geodb_db3dcore_tgz-pJ8-de_uos_igf_db3d_dbms_model3d_SegmentNetBuilderJ8/src/de/uos/igf/db3d/dbms/newModel4d/ServicesFor4DObjects.java

```diff
@@ -0,0 +1,294 @@
+				interpolatedPoints.putAll(TimeStepBuilder
+						.getPointTubesAtInstance(component, date));
+			}
+
+			Map<Integer, Element4D> elements4D = segmentNet
+					.getNetElements(date);
+
+			SegmentElt3D[] elements = new SegmentElt3D[elements4D.size()];
+
+			int cnt = 0;
+
+			for (Integer segmentID : elements4D.keySet()) {
+
+				Segment4D tmp = (Segment4D) elements4D.get(segmentID);
+
+				SegmentElt3D segment = new SegmentElt3D(
+						interpolatedPoints.get(tmp.getIDstart()),
+						interpolatedPoints.get(tmp.getIDend()), sop);
+
+				elements[cnt] = segment;
+				cnt++;
+			}
+
+			// add the component to the SegmentNetBuilder
+			sNB.addComponent(elements);
+
+			builder.setSpatialPart(sNB.getSegmentNet());
+		}
+
+		// create all the Triangel3D objects
+		// for every component:
+
+		// Was benötige ich? PointTubes und Elemente unterteilt nach
+		// Components!
+		// Vermaschung = Elemente kommen von den Netzen! PTs von den
+		// Components!
+
+		// get the right components:
+
+		// now we have the index of the right topology for this date.
+		// so get the right points:
+		// TODO: Do this for every component!
+
+		else if (net.getType() == ComplexGeoObj.TRIANGLE_NET_4D) {
+
+			TriangleNet4D triangleNet = (TriangleNet4D) net;
+
+			Map<Integer, Point3D> interpolatedPoints = new HashMap<Integer, Point3D>();
+
+			for (Component4D component : triangleNet.getValidComponents(date)) {
+				interpolatedPoints.putAll(TimeStepBuilder
+						.getPointTubesAtInstance(component, date));
+			}
+
+			Map<Integer, Element4D> elements4D = triangleNet
+					.getNetElements(date);
+
+			TriangleElt3D[] elements = new TriangleElt3D[elements4D.size()];
+
+			int cnt = 0;
+
+			for (Integer triangleID : elements4D.keySet()) {
+
+				Triangle4D tmp = (Triangle4D) elements4D.get(triangleID);
+
+				TriangleElt3D triangle = new TriangleElt3D(
+						interpolatedPoints.get(tmp.getIDzero()),
+						interpolatedPoints.get(tmp.getIDone()),
+						interpolatedPoints.get(tmp.getIDtwo()), sop);
+
+				elements[cnt] = triangle;
+				cnt++;
+			}
+
+			// add the component to the TriangleNetBuilder
+			tNB.addComponent(elements);
+
+			builder.setSpatialPart(tNB.getTriangleNet());
+		}
+
+		// create all the Tetrahedron3D objects
+		else if (net.getType() == ComplexGeoObj.TETRAHEDRON_NET_4D) {
+
+			TetrahedronNet4D tetrahedronNet = (TetrahedronNet4D) net;
+
+			Map<Integer, Point3D> interpolatedPoints = new HashMap<Integer, Point3D>();
+
+			for (Component4D component : tetrahedronNet
+					.getValidComponents(date)) {
+				interpolatedPoints.putAll(TimeStepBuilder
+						.getPointTubesAtInstance(component, date));
+			}
+
+			Map<Integer, Element4D> elements4D = tetrahedronNet
+					.getNetElements(date);
+
+			TetrahedronElt3D[] elements = new TetrahedronElt3D[elements4D
+					.size()];
+
+			int cnt = 0;
+
+			for (Integer tetrahedronID : elements4D.keySet()) {
+
+				Tetrahedron4D tmp = (Tetrahedron4D) elements4D
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.